File: backend/app/api/v1/entities.py
"""Graph Entity endpoints.

Canonical endpoints per CONTRACT v0.3.2 §9.2.
"""

import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, Query, Response, status
from sqlalchemy.orm import Session

from app.db.session import get_db
from app.schemas.graph import (
    EntityProvenanceResponse,
    GraphResponse,
)
from app.schemas.graph_entity import (
    GraphEntityCreate,
    GraphEntityResponse,
    GraphEntityUpdate,
)
from app.services import graph_service

router = APIRouter(tags=["entities"])


@router.post(
    "/workspaces/{workspace_id}/entities",
    response_model=GraphEntityResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_entity(
    workspace_id: uuid.UUID,
    data: GraphEntityCreate,
    db: Annotated[Session, Depends(get_db)],
) -> GraphEntityResponse:
    """Manually create an entity in a workspace per CONTRACT §9.2."""
    entity = graph_service.create_entity(db, workspace_id, data)
    return GraphEntityResponse.model_validate(entity)


@router.get(
    "/workspaces/{workspace_id}/entities",
    response_model=list[GraphEntityResponse],
    status_code=status.HTTP_200_OK,
)
def list_entities(
    workspace_id: uuid.UUID,
    db: Annotated[Session, Depends(get_db)],
    entity_type: Annotated[str | None, Query()] = None,
    cluster_id: Annotated[uuid.UUID | None, Query()] = None,
    limit: Annotated[int, Query(ge=1, le=1000)] = 500,
) -> list[GraphEntityResponse]:
    """List entities in a workspace."""
    entities = graph_service.list_entities(
        db, workspace_id, entity_type=entity_type, cluster_id=cluster_id, limit=limit
    )
    return [GraphEntityResponse.model_validate(e) for e in entities]


@router.get(
    "/entities/{entity_id}",
    response_model=GraphEntityResponse,
    status_code=status.HTTP_200_OK,
)
def get_entity(
    entity_id: uuid.UUID,
    db: Annotated[Session, Depends(get_db)],
) -> GraphEntityResponse:
    """Retrieve an entity by ID per CONTRACT §9.2."""
    entity = graph_service.get_entity(db, entity_id)
    return GraphEntityResponse.model_validate(entity)


@router.patch(
    "/entities/{entity_id}",
    response_model=GraphEntityResponse,
    status_code=status.HTTP_200_OK,
)
def update_entity(
    entity_id: uuid.UUID,
    data: GraphEntityUpdate,
    db: Annotated[Session, Depends(get_db)],
) -> GraphEntityResponse:
    """Update an entity per CONTRACT §9.2."""
    entity = graph_service.update_entity(db, entity_id, data)
    return GraphEntityResponse.model_validate(entity)


@router.delete(
    "/entities/{entity_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
def delete_entity(
    entity_id: uuid.UUID,
    db: Annotated[Session, Depends(get_db)],
) -> Response:
    """Delete an entity per CONTRACT §9.2."""
    graph_service.delete_entity(db, entity_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.get(
    "/entities/{entity_id}/neighborhood",
    response_model=GraphResponse,
    status_code=status.HTTP_200_OK,
)
def get_entity_neighborhood(
    entity_id: uuid.UUID,
    db: Annotated[Session, Depends(get_db)],
) -> GraphResponse:
    """Return the 1-hop graph neighborhood for an entity per CONTRACT §9.2."""
    return graph_service.get_entity_neighborhood(db, entity_id)


@router.get(
    "/entities/{entity_id}/provenance",
    response_model=EntityProvenanceResponse,
    status_code=status.HTTP_200_OK,
)
def get_entity_provenance(
    entity_id: uuid.UUID,
    db: Annotated[Session, Depends(get_db)],
) -> EntityProvenanceResponse:
    """Return provenance sources for an entity per CONTRACT §9.2."""
    return graph_service.get_entity_provenance(db, entity_id)

File: frontend/src/api/entities.ts
import { apiClient } from './client';
import {
  GraphEntity,
  GraphEntityCreate,
  GraphEntityUpdate,
  EntityProvenanceResponse,
} from '@/types/graph_entity';

export async function createEntity(
  workspaceId: string,
  payload: GraphEntityCreate
): Promise<GraphEntity> {
  return apiClient.post<GraphEntity>(`/workspaces/${workspaceId}/entities`, payload);
}

export async function updateEntity(
  entityId: string,
  payload: GraphEntityUpdate
): Promise<GraphEntity> {
  return apiClient.patch<GraphEntity>(`/entities/${entityId}`, payload);
}

export async function deleteEntity(entityId: string): Promise<void> {
  return apiClient.delete<void>(`/entities/${entityId}`);
}

export async function getEntity(entityId: string): Promise<GraphEntity> {
  return apiClient.get<GraphEntity>(`/entities/${entityId}`);
}

export async function listEntities(
  workspaceId: string,
  params?: { entity_type?: string; cluster_id?: string; limit?: number }
): Promise<GraphEntity[]> {
  const searchParams = new URLSearchParams();
  if (params?.entity_type) searchParams.append('entity_type', params.entity_type);
  if (params?.cluster_id) searchParams.append('cluster_id', params.cluster_id);
  if (params?.limit !== undefined) searchParams.append('limit', params.limit.toString());
  const qs = searchParams.toString();
  return apiClient.get<GraphEntity[]>(
    `/workspaces/${workspaceId}/entities${qs ? `?${qs}` : ''}`
  );
}

export async function getEntityProvenance(
  entityId: string
): Promise<EntityProvenanceResponse> {
  return apiClient.get<EntityProvenanceResponse>(`/entities/${entityId}/provenance`);
}
